chore: backend service alb https support

dannyrandall opened the pull request:
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

dannyrandall committed:
pass `httpsEnabled` and `certEnabled` to backend service template

dannyrandall committed:
Merge branch 'mainline' into chore/backend-https

dannyrandall committed:
fixed some test cases

dannyrandall committed:
https path alias test case

dannyrandall committed:
validate backend alb before deploy

dannyrandall committed:
remove todo

iamhopaul123 commented on internal/pkg/cli/deploy/deploy.go:
```diff
@@ -1197,35 +1229,47 @@ func validateAppVersionForAlias(appName string, appVersionGetter versionGetter)
 			return &errSvcWithNoALBAliasDeployingToEnvWithImportedCerts{
-				name:    d.name,
-				envName: d.env.Name,
+				name:    in.ServiceName,
+				envName: in.Env.Name,
```
Blame me but there's seem to be a bug. When http is disabled (or not set for backend svc)

```yaml
http: false
```

We shouldn't error out saying if any cert is imported, `alias` has to be set.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -29,36 +30,53 @@ type backendSvcReadParser interface {
+
+	if conf.Env.HasImportedCerts() {
+		b.certImported = true
+		b.httpsEnabled = true
```
This shouldn't be true if `http` is disabled or empty.

dannyrandall committed:
Merge branch 'mainline' into chore/backend-https

efekarakus commented on internal/pkg/cli/deploy/deploy.go:
```diff
@@ -1197,35 +1229,47 @@ func validateAppVersionForAlias(appName string, appVersionGetter versionGetter)
-		if err := validateAppVersionForAlias(d.app.Name, d.appVersionGetter); err != nil {
-			logAppVersionOutdatedError(aws.StringValue(d.lbMft.Name))
+
+	if in.App.Domain != "" {
```
Similar to boss's comment, the app's domain I think isn't relevant for backend service's `alias`.

At least for now, we'll start of requiring only [this if-statement](https://github.com/aws/copilot-cli/pull/3583/files?diff=split&w=0#diff-d23f1b40938f2960e3b84a2e5799f6e56df9805461923e4319fe19bb22e87763R1252) where if alias is set then they must have imported certs for backend services

efekarakus reviewed (commented):
+1 to all of Penghao's feedback :)

dannyrandall committed:
fix `httpsEnabled` if alb is disabled

dannyrandall committed:
addr fb and add testing:
- don't fail alb validation if alb is disabled
- don't check alias against app domain
- don't require alias if env has certs

dannyrandall committed:
remove extra space added in lbws function

dannyrandall committed:
fix https test case

dannyrandall committed:
fix test case

efekarakus commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -29,36 +30,55 @@ type backendSvcReadParser interface {
+
+	if conf.Env.HasImportedCerts() {
+		b.certImported = true
+		b.httpsEnabled = b.albEnabled
```
nice

iamhopaul123 reviewed (approved):
🚢

mergify[bot] committed:
Merge branch 'mainline' into chore/backend-https
